core/src/main/java/org/apache/iceberg/SnapshotManager.java

```diff
@@ -0,0 +1,151 @@
+    this.table = table;
+  }
+
+  @Override
+  protected ManageSnapshots self() {
+    return this;
+  }
+
+  @Override
+  protected String operation() {
+    Snapshot manageSnapshot = base.snapshot(targetSnapshotId);
+    return manageSnapshot.operation();
+  }
+
+  @Override
+  public ManageSnapshots cherrypick(long snapshotId) {
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null,
+        "Cannot cherry pick unknown snapshot id: %s", snapshotId);
+
+    operation = SnapshotManagerOperation.CHERRYPICK;
+    this.targetSnapshotId = snapshotId;
+
+    // Pick modifications from the snapshot
+    Snapshot cherryPickSnapshot = base.snapshot(this.targetSnapshotId);
+    // only append operations are currently supported
+    if (!cherryPickSnapshot.operation().equals(DataOperations.APPEND)) {
+      throw new UnsupportedOperationException("Can cherry pick only append operations");
+    }
+
+    for (DataFile addedFile : cherryPickSnapshot.addedFiles()) {
+      add(addedFile);
+    }
+    // this property is set on target snapshot that will get published
+    String wapId = stagedWapId(cherryPickSnapshot);
+    boolean isWapWorkflow =  wapId != null && !wapId.isEmpty();
+    if (isWapWorkflow) {
+      set(SnapshotSummary.PUBLISHED_WAP_ID_PROP, wapId);
+    }
+
+    return this;
+  }
+
+  @Override
+  public ManageSnapshots setCurrentSnapshot(long snapshotId) {
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null,
+        "Cannot roll back to unknown snapshot id: %s", snapshotId);
+
+    operation = SnapshotManagerOperation.ROLLBACK;
+    this.targetSnapshotId = snapshotId;
+
+    return this;
+  }
+
+  @Override
+  public ManageSnapshots rollbackToTime(long timestampMillis) {
+    operation = SnapshotManagerOperation.ROLLBACK;
+    // find the latest snapshot by timestamp older than timestampMillis
+    Snapshot snapshot = SnapshotUtil.findLatestSnapshotOlderThan(table, timestampMillis);
+    Preconditions.checkArgument(snapshot != null,
+        "Cannot roll back, no valid snapshot older than: %s", timestampMillis);
+    this.targetSnapshotId = snapshot.snapshotId();
+
+    return this;
+  }
+
+  @Override
+  public ManageSnapshots rollbackTo(long snapshotId) {
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null,
+        "Cannot roll back to unknown snapshot id: %s", snapshotId);
+
+    ValidationException.check(SnapshotUtil.isCurrentAncestor(table, snapshotId),
+        "Cannot roll back to snapshot, not an ancestor of the current state: %s", snapshotId);
+    return setCurrentSnapshot(snapshotId);
+  }
+
+  @Override
+  public Snapshot apply() {
+    // common checks
+    ValidationException.check(operation != null,
+        "Need to define operation on snapshot: call setCurrentSnapshot, rollbackAtTime or cherrypick");
+
+    if (targetSnapshotId == null) {
+      // if no target snapshot was configured then NOOP by returning current state
+      return table.currentSnapshot();
```

Comment:
This logic seems to conflict with the validation above. If it's okay to not have a target snapshot ID and the result is a noop, then we don't need the check that there is an operation, right?

Reply:
done.